src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs

```diff
@@ -218,7 +167,128 @@ private async Task UpdateJobRecordAsync(CancellationToken cancellationToken)
+            string currentBatchId = batchIdPrefix + "-" + progress.Page.ToString("d6");
+
+            List<Tuple<string, string>> queryParametersList = new List<Tuple<string, string>>(sharedQueryParametersList);
+            if (progress.ContinuationToken != null)
+            {
+                queryParametersList.Add(Tuple.Create(KnownQueryParameterNames.ContinuationToken, progress.ContinuationToken));
+            }
+
+            // Process the export if:
+            // 1. There is continuation token, which means there is more resource to be exported.
+            // 2. There is no continuation token but the page is 0, which means it's the initial export.
+            while (progress.ContinuationToken != null || progress.Page == 0)
+            {
+                SearchResult searchResult = null;
+
+                // Search and process the results.
+                using (IScoped<ISearchService> searchService = _searchServiceFactory())
+                {
+                    if (compartmentSearch)
+                    {
+                        searchResult = await searchService.Value.SearchCompartmentAsync(
+                            KnownResourceTypes.Patient,
+                            progress.TriggeringResourceId,
+                            null,
+                            queryParametersList,
+                            cancellationToken);
+                    }
+                    else
+                    {
+                        switch (_exportJobRecord.ExportType)
+                        {
+                            case ExportJobType.All:
+                                searchResult = await searchService.Value.SearchAsync(
+                                    null,
+                                    queryParametersList,
+                                    cancellationToken);
+                                break;
+                            case ExportJobType.Patient:
+                                searchResult = await searchService.Value.SearchAsync(
+                                    KnownResourceTypes.Patient,
+                                    queryParametersList,
+                                    cancellationToken);
+                                break;
+                        }
+                    }
+                }
+
+                if (_exportJobRecord.ExportType == ExportJobType.Patient && !compartmentSearch)
+                {
+                    uint resultIndex = 0;
+                    foreach (SearchResultEntry result in searchResult.Results)
+                    {
+                        // If a job is resumed in the middle of processing patient compartment resources it will skip patients it has already exported compartment information for.
+                        if (progress.SubSearch != null && result.Resource.ResourceId != progress.SubSearch.TriggeringResourceId)
+                        {
+                            continue;
```

Comment:
In scenarios where we hit this code, the value of `resultIndex` will be out-of sync with the iterator. We could end up overwriting existing data in the blobs.